torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2386,6 +2386,41 @@ Tensor linalg_eig_backward(const std::vector<torch::autograd::Variable> &grads,
   }
 }
 
+// jvp functions for eigenvalues and eigenvectors are separate
+// because currently forward AD only works with one rule per output
+Tensor eigh_jvp_eigenvalues(
+    const Tensor& input_tangent,
+    const Tensor& eigenvalues,
+    const Tensor& eigenvectors) {
+  // An extended collection of matrix derivative results for forward and reverse mode automatic differentiation
+  // https://ora.ox.ac.uk/objects/uuid:8d0c0a29-c92b-4153-a1d2-38b276e93124
+  // Section 3.1 Eigenvalues and eigenvectors
+
+  auto tmp = at::matmul(at::matmul(eigenvectors.transpose(-2, -1).conj(), input_tangent), eigenvectors);
```

Comment:
I believe that the input tangent needs to be symmetrised as: `const auto in = 0.5 * (input_tangent + input_tangent.transpose(-2, -1).conj())`.

Denote `Her(n)` the `n x n` Hermitian matrices and `U(n)` the unitary matrices. We have that `eigh : Her(n) -> R^n x U(n)`. This means that the differential of `eigh` goes from the tangent to `Her(n)` to the tangent of `R^n` times the tangent of `U(n)` (at the output matrices). Now, the tangent of `Her(n)` is `Her(n)` itself as  `Her(n)` is just a linear subspace. As such, the input needs to be symmetric for this function to make sense.

Computing `(A + A^H)/2` happens to be the orthogonal projection of a matrix onto the space `Her(n)`. The theorem that formalises all this is the one that says that the differential of a map on an embedded manifold is the differential of the map on the total space restricted to the tangent space of the embedded manifold.

I wrote all this down in a comment in `eigh_backward`.